packages/hardhat-launcher/src/completion.ts

```diff
@@ -0,0 +1,139 @@
+
+export async function main() {
+  const cmd = process.argv[2];
+
+  if (cmd === "install") {
+    await tabtab
+      .install({
+        name: "hh",
+        completer: "hh-completion",
+      })
+      .catch((err: any) => {
+        console.error(
+          "There was a problem installing Hardhat's completion",
+          err
+        );
+      });
+
+    return;
+  }
+
+  if (cmd === "completion") {
+    let pathToHardhatPackageJson: string;
+
+    try {
+      // tslint:disable-next-line no-implicit-dependencies
+      pathToHardhatPackageJson = require.resolve("hardhat/package.json", {
+        paths: [process.cwd()],
+      });
+    } catch (e) {
+      // not inside a hardhat project
+      return;
+    }
+
+    try {
+      const env = tabtab.parseEnv(process.env);
+
+      // check hh's dependency on hardhat
+      const hardhatVersion = require(pathToHardhatPackageJson).version;
+      const pathToHardhatAutocomplete = getRequirePathFromCwd(
+        "hardhat/internal/cli/autocomplete"
+      );
+
+      if (
+        !semver.satisfies(hardhatVersion, REQUIRED_HARDHAT_VERSION_RANGE) ||
+        pathToHardhatAutocomplete === null
+      ) {
+        logWarningWithThrottling(
+          `Couldn't get autocomplete for this project. The installed version of hh requires a hardhat version that satisfies ${REQUIRED_HARDHAT_VERSION_RANGE}, but this project uses ${hardhatVersion}`
+        );
+        return;
+      }
+
+      // check hardhat's dependency on hh
+      const hhVersion = require("../../package.json").version;
+
+      const {
+        complete,
+        HARDHAT_COMPLETE_FILES,
+        REQUIRED_HH_VERSION_RANGE,
+      } = require(pathToHardhatAutocomplete);
+
+      if (!semver.satisfies(hhVersion, REQUIRED_HH_VERSION_RANGE)) {
+        logWarningWithThrottling(
+          `Couldn't get autocomplete for this project. The version of hardhat used in this project requires an hh version that satisfies ${REQUIRED_HH_VERSION_RANGE}, but your version of hh is ${hhVersion}`
+        );
+        return;
+      }
+
+      // get and print suggestions
+      const suggestions = await complete(env);
+
+      if (Array.isArray(suggestions)) {
+        return tabtab.log(suggestions);
+      }
+
+      if (suggestions === HARDHAT_COMPLETE_FILES) {
+        return tabtab.logFiles();
+      }
+
+      console.error("Couldn't complete the command, please report this issue");
+      return tabtab.log([]);
+    } catch (e) {
+      log(e.message);
+      return tabtab.log([]);
+    }
+  }
+
+  console.error(
+    `Unrecognized command "${cmd}". You can install Hardhat completion with the "install" command.`
+  );
+  process.exit(1);
+}
+
+function logWarningWithThrottling(message: string) {
+  // we need to write a file to disk to do the throttling because zsh calls the
+  // autocomplete function several times, and these are separate processes
+  const throttleFile = path.join(__dirname, ".hh-throttle-file");
```

Comment:
This is saving this into a `node_modules`. I think it can lead to problems.

Reply:
maybe `process.cwd()` instead of `__dirname`? This errors shouldn't be common.

Reply:
I think using `process.cwd()` will mean that this file can be created in (multiple) places of the user's repository, which seems annoying (I would find it annoying at least, even if I can .gitignore it).

What problems can we run into if we use `node_modules`?

Reply:
It wouldn't work with yarn2, it wouldn't work for the people that incorrectly use sudo to install global packages, and there's probably more issues.

I wouldn't assume that there's a global node modules nor that it's writable.

Reply:
We can do something similar to hardhat's global dirs